### PR TITLE
AVRO-4007: [rust] Faster `is_nullable` for UnionSchema

### PR DESCRIPTION
**Pull request by JohnEmhoff**
AVRO-4007

Apologies for the lack of a Jira issue; I'll link an issue as soon as my account request is approved.

## What is the purpose of the change

Writing large amounts of avro data in rust is slow because (in my case) ~40% of total run time is spent in the function `UnionSchema::is_nullable`. The issue is that the `x == Schema::Null` invokes schema canonicalization which is
apparently somewhat slow. I've modified the method to use `match` instead and see a considerable performance improvement.

## Verifying this change

This change does not change any behavior and should be covered by existing tests

## Documentation

- Does this pull request introduce a new feature? no

### CODE REVIEW

**Comment by JohnEmhoff** [comment]
Hmmm, I just realized this will return false for cases like `Union(Int, Union(Null))` if the schema is not canonicalized upfront. I'm not familiar enough with the library to know if this is the case or not, but the performance win is big enough that I'm happy to work through whatever the potential issues are.

File: lang/rust/avro/src/schema.rs [hunk 1]
```diff
@@ -902,7 +902,10 @@ impl UnionSchema {
 
     /// Returns true if the any of the variants of this `UnionSchema` is `Null`.
     pub fn is_nullable(&self) -> bool {
-        !self.schemas.is_empty() && self.schemas.iter().any(|s| s == &Schema::Null)
+        self.schemas.iter().any(|x| match x {
+            Schema::Null => true,
+            _ => false
```
Comment (martin-g) [inline lang/rust/avro/src/schema.rs:907]:
```suggestion
            _ => false,
```

**Comment by martin-g** [comment]
I've approved your account a few days ago but to be able to merge your improvement I also created AVRO-4007 for you!

**Comment by martin-g** [comment]
Thank you, @JohnEmhoff !

**Comment by JohnEmhoff** [comment]
Thank you very much @martin-g ❤️

**Comment by martin-g** [comment]
> The issue is that the `x == Schema::Null` invokes schema canonicalization which is
> apparently somewhat slow.

BTW in current `main` branch (0.17) the schema equality no more uses the canonical forms, so it must be much faster!
See https://github.com/apache/avro/pull/2739

**Pull request merged**